### PR TITLE
[SPARK-10614][core] Add monotonic time to Clock interface.

### PR DESCRIPTION
**Pull request by vanzin**
This change adds a new method to the Clock interface that returns
the time from a monotonic time source, so that code that needs that
feature can also mock the Clock in tests.

The original getTimeMillis and waitTillTime methods are unchanged, since
streaming code that uses the Clock interface seems to rely on wall clock
semantics, not monotonic clock. So, in a way, this doesn't directly
address the problem raised in the bug, that waitTillTime can be affected
by drift, but then the places being modified to use the new API don't
really rely on that API.

The dynamic allocation code was modified to use the new time source,
since they shouldn't be basing their decisions on wall clock time.

For a longer discussion on how monotonic clocks work on Linux/x64, the
following blog post (and links within) shed a lot of light on the safety of
`System.nanoTime()`:

http://btorpey.github.io/blog/2014/02/18/clock-sources-in-linux/

Tested with unit test and also running apps with dynamic allocation on.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 1]
```diff
@@ -288,7 +288,7 @@ private[spark] class ExecutorAllocationManager(
     }
 
     // Update executor target number only after initializing flag is unset
-    updateAndSyncNumExecutorsTarget(clock.getTimeMillis())
+    updateAndSyncNumExecutorsTarget(clock.nanoTime())
```
Comment (MaxGekk) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: System.nanoTime() is not monotonic across CPU cores. Are you sure that this value of `clock.nanoTime()` is not compared to `System.nanoTime()` from another thread? For instance, from `onSchedulerBacklogged()`.

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Are you sure about that?

This is what the javadocs say:

    The same origin is used by all invocations of this method in an instance of a Java virtual
    machine; other virtual machine instances are likely to use a different origin. 

(Comment about "other virtual machines" aside.)

That seems to be corroborated by other sources, too (e.g. https://stackoverflow.com/questions/510462/is-system-nanotime-completely-useless). Seems in a long gone past there might have been issues, but not today.

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Also, just wanted to point out that it's ok if `nanoTime` is not completely monotonic, as long as it's better than `getTimeMillis()`. After all, this code was working with a non-monotonic clock before.

The only problem is if different threads can have wildly different views of what `nanoTime` returns, so that comparing the values makes no sense. I don't believe that to be the case.

Comment (MaxGekk) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: `System.nanoTime()` can use `clock_gettime(CLOCK_MONOTONIC, ...)` on Linux: http://hg.openjdk.java.net/jdk8/jdk8/hotspot/file/tip/src/os/linux/vm/os_linux.cpp#l1453
`CLOCK_MONOTONIC` is based on the Time Stamp Counter (https://en.wikipedia.org/wiki/Time_Stamp_Counter). Its synchronization across CPU cores depends on CPU architecture, for example: _"AMD processors up to the K8 core always incremented the time-stamp counter every clock cycle.[6] Thus, power management features were able to change the number of increments per second, and the values could get out of sync between different cores or processors in the same system."_

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: That is interesting, but according to this:

    http://btorpey.github.io/blog/2014/02/18/clock-sources-in-linux/

`CLOCK_MONOTONIC` uses RDTSC (or RDTSCP when available), which solves that problem. The preferred one (RDTSCP) seems to be available in all modern CPUs. So, `CLOCK_MONOTONIC` actually fulfills the requirements of the `nanoTime()` API (that the source of time is consistent).

(The AMD K8 mentioned in your comment, which came out more than 10 years ago, is where RDTSCP was added.)

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Probably worth to link directly to the following also (which is linked from the article above)

https://stackoverflow.com/questions/10921210/cpu-tsc-fetch-operation-especially-in-multicore-multi-processor-environment

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/util/Clock.scala [hunk 2]
```diff
@@ -21,7 +21,14 @@ package org.apache.spark.util
  * An interface to represent clocks, so that they can be mocked out in unit tests.
  */
 private[spark] trait Clock {
+  /** @return Current system time, in ms. */
   def getTimeMillis(): Long
+  /** @return Current value of monotonic time source, in ns. */
+  def nanoTime(): Long
+  /**
```
Comment (advancedxy) [inline core/src/main/scala/org/apache/spark/util/Clock.scala]: Nit: add blank line between methods?

File: core/src/main/scala/org/apache/spark/util/Clock.scala [hunk 3]
```diff
@@ -36,19 +43,23 @@ private[spark] class SystemClock extends Clock {
...
-  def waitTillTime(targetTime: Long): Long = {
-    var currentTime = 0L
-    currentTime = System.currentTimeMillis()
+  override def waitTillTime(targetTime: Long): Long = {
+    var currentTime = System.currentTimeMillis()
 
     var waitTime = targetTime - currentTime
     if (waitTime <= 0) {
-      return currentTime
+      return getTimeMillis()
```
Comment (advancedxy) [inline core/src/main/scala/org/apache/spark/util/Clock.scala]: Why not just `return currentTime`?

File: core/src/main/scala/org/apache/spark/util/Clock.scala (hunk 2 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/util/Clock.scala]: Old code didn't have them, but sure.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala (hunk 1 above)
Comment (squito) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: this is a good discussion -- its really easy to get confused about the details here, especially as some googling can take you to old info.  Worth including a bit of this on the doc for `SystemClock.nanoTime()`?  Maybe at least just noting that this is dependable across cores as long as its Java7+.

**Review by squito — APPROVED** [review]
lgtm, just a request for a comment

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: +1 for @squito 's comment. Also, could you summarize the above into the PR description a little bit please, @vanzin ?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (except @squito 's comment). Thanks!

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for updating, @vanzin . Thank you, @squito , @MaxGekk , @advancedxy .
Merged to master.
